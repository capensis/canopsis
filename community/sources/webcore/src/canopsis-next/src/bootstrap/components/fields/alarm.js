export { default as CAlarmInfosAttributeField } from '@/components/forms/fields/alarm/c-alarm-infos-attribute-field.vue';
export { default as CAlarmTagField } from '@/components/forms/fields/alarm/c-alarm-tag-field.vue';
export { default as CAlarmField } from '@/components/forms/fields/alarm/c-alarm-field.vue';
