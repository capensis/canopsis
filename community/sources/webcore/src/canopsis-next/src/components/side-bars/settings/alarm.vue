<template lang="pug">
  div
    v-list.pt-0(expand)
      field-title(v-model="settings.widget.title", :title="$t('common.title')")
      v-divider
      field-periodic-refresh(v-model="settings.widget.parameters.periodic_refresh")
      v-divider
      v-list-group(data-test="advancedSettings")
        v-list-tile(slot="activator") {{ $t('settings.advancedSettings') }}
        v-list.grey.lighten-4.px-2.py-0(expand)
          field-default-sort-column(
            v-model="settings.widget.parameters.sort",
            :columns="settings.widget.parameters.widgetColumns",
            :columns-label="$t('settings.columnName')"
          )
          v-divider
          field-columns(
            v-model="settings.widget.parameters.widgetColumns",
            :label="$t('settings.columnNames')",
            with-html,
            with-color-indicator
          )
          v-divider
          field-columns(
            v-model="settings.widget.parameters.widgetGroupColumns",
            :label="$t('settings.groupColumnNames')",
            with-html,
            with-color-indicator
          )
          v-divider
          field-columns(
            v-model="settings.widget.parameters.serviceDependenciesColumns",
            :label="$t('settings.trackColumnNames')",
            with-color-indicator
          )
          v-divider
          field-default-elements-per-page(v-model="settings.userPreferenceContent.itemsPerPage")
          v-divider
          field-opened-resolved-filter(v-model="settings.widget.parameters.opened")
          v-divider
          template(v-if="hasAccessToListFilters")
            field-filters(
              v-model="settings.widget.parameters.mainFilter",
              :entities-type="$constants.ENTITIES_TYPES.alarm",
              :filters.sync="settings.widget.parameters.viewFilters",
              :condition.sync="settings.widget.parameters.mainFilterCondition",
              :has-access-to-add-filter="hasAccessToAddFilter",
              :has-access-to-edit-filter="hasAccessToEditFilter",
              @input="updateMainFilterUpdatedAt"
            )
            v-divider
          template(v-if="hasAccessToListRemediationInstructionsFilters")
            field-remediation-instructions-filters(
              v-model="settings.widget.parameters.remediationInstructionsFilters",
              :has-access-to-add-filter="hasAccessToAddRemediationInstructionsFilter",
              :has-access-to-edit-filter="hasAccessToEditRemediationInstructionsFilter"
            )
            v-divider
          field-live-reporting(v-model="settings.widget.parameters.liveReporting")
          v-divider
          field-info-popup(
            v-model="settings.widget.parameters.infoPopups",
            :columns="settings.widget.parameters.widgetColumns"
          )
          v-divider
          field-text-editor(
            data-test="widgetMoreInfoTemplate",
            v-model="settings.widget.parameters.moreInfoTemplate",
            :title="$t('settings.moreInfosModal')"
          )
          v-divider
          field-grid-range-size(
            v-model="settings.widget.parameters.expandGridRangeSize",
            :title="$t('settings.expandGridRangeSize')"
          )
          v-divider
          field-switcher(
            data-test="isHtmlEnabledOnTimeLine",
            v-model="settings.widget.parameters.isHtmlEnabledOnTimeLine",
            :title="$t('settings.isHtmlEnabledOnTimeLine')"
          )
          v-divider
          v-list-group(data-test="ackGroup")
            v-list-tile(slot="activator") Ack
            v-list.grey.lighten-4.px-2.py-0(expand)
            field-switcher(
              data-test="isAckNoteRequired",
              v-model="settings.widget.parameters.isAckNoteRequired",
              :title="$t('settings.isAckNoteRequired')"
            )
            v-divider
            field-switcher(
              data-test="isMultiAckEnabled",
              v-model="settings.widget.parameters.isMultiAckEnabled",
              :title="$t('settings.isMultiAckEnabled')"
            )
            v-divider
            field-fast-ack-output(v-model="settings.widget.parameters.fastAckOutput")
          v-divider
          field-switcher(
            v-model="settings.widget.parameters.isSnoozeNoteRequired",
            :title="$t('settings.isSnoozeNoteRequired')"
          )
          v-divider
          field-enabled-limit(
            v-model="settings.widget.parameters.linksCategoriesAsList",
            :title="$t('settings.linksCategoriesAsList')",
            :label="$t('settings.linksCategoriesLimit')"
          )
          v-divider
          export-csv-form(v-model="settings.widget.parameters", datetime-format)
          v-divider
          field-switcher(
            v-model="settings.widget.parameters.sticky_header",
            :title="$t('settings.stickyHeader')"
          )
      v-divider
    v-btn.primary(data-test="submitAlarms", @click="submit") {{ $t('common.save') }}
</template>

<script>
import { get } from 'lodash';

import { PAGINATION_LIMIT } from '@/config';
import { SIDE_BARS } from '@/constants';

import { alarmListWidgetToForm, formToAlarmListWidget } from '@/helpers/forms/widgets/alarm';

import { widgetSettingsMixin } from '@/mixins/widget/settings';
import { permissionsWidgetsAlarmsListFilters } from '@/mixins/permissions/widgets/alarms-list/filters';
import { permissionsWidgetsAlarmsListRemediationInstructionsFilters }
  from '@/mixins/permissions/widgets/alarms-list/remediation-instructions-filters';

import FieldTitle from './fields/common/title.vue';
import FieldDefaultSortColumn from './fields/common/default-sort-column.vue';
import FieldColumns from './fields/common/columns.vue';
import FieldLiveReporting from './fields/common/live-reporting.vue';
import FieldPeriodicRefresh from './fields/common/periodic-refresh.vue';
import FieldDefaultElementsPerPage from './fields/common/default-elements-per-page.vue';
import FieldFilters from './fields/common/filters.vue';
import FieldTextEditor from './fields/common/text-editor.vue';
import FieldSwitcher from './fields/common/switcher.vue';
import FieldFastAckOutput from './fields/alarm/fast-ack-output.vue';
import FieldGridRangeSize from './fields/common/grid-range-size.vue';
import FieldRemediationInstructionsFilters from './fields/common/remediation-instructions-filters.vue';
import FieldOpenedResolvedFilter from './fields/alarm/opened-resolved-filter.vue';
import FieldInfoPopup from './fields/alarm/info-popup.vue';
import FieldEnabledLimit from './fields/common/enabled-limit.vue';
import ExportCsvForm from './forms/export-csv.vue';

/**
 * Component to regroup the alarms list settings fields
 */
export default {
  name: SIDE_BARS.alarmSettings,
  $_veeValidate: {
    validator: 'new',
  },
  components: {
    FieldTitle,
    FieldDefaultSortColumn,
    FieldColumns,
    FieldLiveReporting,
    FieldPeriodicRefresh,
    FieldDefaultElementsPerPage,
    FieldOpenedResolvedFilter,
    FieldFilters,
    FieldTextEditor,
    FieldSwitcher,
    FieldFastAckOutput,
    FieldGridRangeSize,
    FieldRemediationInstructionsFilters,
    FieldInfoPopup,
    FieldEnabledLimit,
    ExportCsvForm,
  },
  mixins: [
    widgetSettingsMixin,
    permissionsWidgetsAlarmsListFilters,
    permissionsWidgetsAlarmsListRemediationInstructionsFilters,
  ],
  data() {
    const { widget } = this.config;

    return {
      settings: {
        widget: alarmListWidgetToForm(widget),
        userPreferenceContent: { itemsPerPage: PAGINATION_LIMIT },
      },
    };
  },
  mounted() {
    const { content } = this.userPreference;

    this.settings.userPreferenceContent = {
      itemsPerPage: get(content, 'itemsPerPage', PAGINATION_LIMIT),
    };
  },
  methods: {
    updateMainFilterUpdatedAt() {
      this.settings.widget.parameters.mainFilterUpdatedAt = Date.now();
    },

    prepareWidgetSettings() {
      const { widget } = this.settings;

      return formToAlarmListWidget(widget);
    },
  },
};
</script>
