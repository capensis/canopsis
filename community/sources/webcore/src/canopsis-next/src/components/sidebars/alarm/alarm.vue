<template lang="pug">
  widget-settings(:submitting="submitting", @submit="submit")
    field-title(v-model="form.title")
    v-divider
    field-periodic-refresh(v-model="form.parameters.periodic_refresh")
    v-divider
    widget-settings-group(:title="$t('settings.advancedSettings')")
      field-default-sort-column(
        v-model="form.parameters.sort",
        :columns="sortablePreparedWidgetColumns",
        :columns-label="$t('settings.columnName')"
      )
      v-divider
      field-columns(
        v-model="form.parameters.widgetColumns",
        :template="form.parameters.widgetColumnsTemplate",
        :templates="alarmColumnsWidgetTemplates",
        :templates-pending="widgetTemplatesPending",
        :label="$t('settings.columnNames')",
        :type="$constants.ENTITIES_TYPES.alarm",
        with-template,
        with-html,
        with-color-indicator,
        @update:template="updateWidgetColumnsTemplate"
      )
      v-divider
      field-resize-column-behavior(v-model="form.parameters.columns")
      v-divider
      field-columns(
        v-model="form.parameters.widgetGroupColumns",
        :template="form.parameters.widgetGroupColumnsTemplate",
        :templates="alarmColumnsWidgetTemplates",
        :templates-pending="widgetTemplatesPending",
        :label="$t('settings.groupColumnNames')",
        :type="$constants.ENTITIES_TYPES.alarm",
        with-html,
        with-color-indicator,
        @update:template="updateWidgetGroupColumnsTemplate"
      )
      v-divider
      field-columns(
        v-model="form.parameters.serviceDependenciesColumns",
        :template="form.parameters.serviceDependenciesColumnsTemplate",
        :templates="entityColumnsWidgetTemplates",
        :templates-pending="widgetTemplatesPending",
        :label="$t('settings.trackColumnNames')",
        :type="$constants.ENTITIES_TYPES.entity",
        with-color-indicator,
        @update:template="updateServiceDependenciesColumnsTemplate"
      )
      v-divider
      field-default-elements-per-page(v-model="form.parameters.itemsPerPage")
      v-divider
      field-density(v-model="form.parameters.dense")
      v-divider
      field-opened-resolved-filter(v-model="form.parameters.opened")
      v-divider
      template(v-if="hasAccessToListFilters")
        field-filters(
          v-model="form.parameters.mainFilter",
          :filters.sync="form.filters",
          :widget-id="widget._id",
          :addable="hasAccessToAddFilter",
          :editable="hasAccessToEditFilter",
          with-alarm,
          with-entity,
          with-pbehavior
        )
        v-divider
      template(v-if="hasAccessToListRemediationInstructionsFilters")
        field-remediation-instructions-filters(
          v-model="form.parameters.remediationInstructionsFilters",
          :addable="hasAccessToAddRemediationInstructionsFilter",
          :editable="hasAccessToEditRemediationInstructionsFilter"
        )
        v-divider
      field-switcher(
        v-model="form.parameters.clearFilterDisabled",
        :title="$t('settings.clearFilterDisabled')"
      )
      v-divider
      field-live-reporting(v-model="form.parameters.liveReporting")
      v-divider
      field-info-popup(
        v-model="form.parameters.infoPopups",
        :columns="preparedWidgetColumns"
      )
      v-divider
      field-text-editor-with-template(
        :value="form.parameters.moreInfoTemplate",
        :template="form.parameters.moreInfoTemplateTemplate",
        :title="$t('settings.moreInfosModal')",
        :variables="alarmVariables",
        :templates="alarmMoreInfosWidgetTemplates",
        addable,
        removable,
        @input="updateMoreInfo"
      )
      v-divider
      field-grid-range-size(
        v-model="form.parameters.expandGridRangeSize",
        :title="$t('settings.expandGridRangeSize')"
      )
      v-divider
      field-switcher(
        v-model="form.parameters.isHtmlEnabledOnTimeLine",
        :title="$t('settings.isHtmlEnabledOnTimeLine')"
      )
      v-divider
      widget-settings-group(:title="$t('common.ack')")
        field-switcher(
          v-model="form.parameters.isAckNoteRequired",
          :title="$t('settings.isAckNoteRequired')"
        )
        v-divider
        field-switcher(
          v-model="form.parameters.isMultiAckEnabled",
          :title="$t('settings.isMultiAckEnabled')"
        )
        v-divider
        field-fast-action-output(
          v-model="form.parameters.fastAckOutput",
          :label="$t('settings.fastAckOutput')"
        )
      v-divider
      widget-settings-group(:title="$t('common.cancel')")
        field-fast-action-output(
          v-model="form.parameters.fastCancelOutput",
          :label="$t('settings.fastCancelOutput')"
        )
      v-divider
      field-switcher(
        v-model="form.parameters.isSnoozeNoteRequired",
        :title="$t('settings.isSnoozeNoteRequired')"
      )
      v-divider
      field-switcher(
        v-model="form.parameters.isMultiDeclareTicketEnabled",
        :title="$t('settings.isMultiDeclareTicketEnabled')"
      )
      v-divider
      field-number(
        v-model="form.parameters.inlineLinksCount",
        :title="$t('settings.inlineLinksCount')"
      )
      v-divider
      export-csv-form(
        v-model="form.parameters",
        :type="$constants.ENTITIES_TYPES.alarm",
        :templates="alarmColumnsWidgetTemplates",
        :templates-pending="widgetTemplatesPending",
        datetime-format
      )
      v-divider
      field-switcher(
        v-model="form.parameters.sticky_header",
        :title="$t('settings.stickyHeader')"
      )
      v-divider
      widget-settings-group(:title="$t('settings.kioskMode')")
        field-switcher(
          v-model="form.parameters.kiosk.hideActions",
          :title="$t('settings.kiosk.hideActions')"
        )
        v-divider
        field-switcher(
          v-model="form.parameters.kiosk.hideMassSelection",
          :title="$t('settings.kiosk.hideMassSelection')"
        )
        v-divider
        field-switcher(
          v-model="form.parameters.kiosk.hideToolbar",
          :title="$t('settings.kiosk.hideToolbar')"
        )
    v-divider
    charts-form(v-model="form.parameters.charts")
    v-divider
</template>

<script>
import { SIDE_BARS, ALARM_UNSORTABLE_FIELDS, ALARM_FIELDS_TO_LABELS_KEYS } from '@/constants';

import { formToWidgetColumns } from '@/helpers/entities/widget/column/form';
import { getWidgetColumnLabel, getWidgetColumnSortable } from '@/helpers/entities/widget/list';

import { widgetSettingsMixin } from '@/mixins/widget/settings';
import { entitiesInfosMixin } from '@/mixins/entities/infos';
import { alarmVariablesMixin } from '@/mixins/widget/variables';
import { widgetTemplatesMixin } from '@/mixins/widget/templates';
import { permissionsWidgetsAlarmsListFilters } from '@/mixins/permissions/widgets/alarms-list/filters';
import { permissionsWidgetsAlarmsListRemediationInstructionsFilters }
  from '@/mixins/permissions/widgets/alarms-list/remediation-instructions-filters';

import WidgetSettingsGroup from '../partials/widget-settings-group.vue';
import WidgetSettings from '../partials/widget-settings.vue';
import FieldTitle from '../form/fields/title.vue';
import FieldDefaultSortColumn from '../form/fields/default-sort-column.vue';
import FieldColumns from '../form/fields/columns.vue';
import FieldPeriodicRefresh from '../form/fields/periodic-refresh.vue';
import FieldDefaultElementsPerPage from '../form/fields/default-elements-per-page.vue';
import FieldFilters from '../form/fields/filters.vue';
import FieldTextEditorWithTemplate from '../form/fields/text-editor-with-template.vue';
import FieldSwitcher from '../form/fields/switcher.vue';
import FieldNumber from '../form/fields/number.vue';
import ExportCsvForm from '../form/export-csv.vue';
import ChartsForm from '../chart/form/charts-form.vue';

import FieldGridRangeSize from './form/fields/grid-range-size.vue';
import FieldRemediationInstructionsFilters from './form/fields/remediation-instructions-filters.vue';
import FieldDensity from './form/fields/density.vue';
import FieldLiveReporting from './form/fields/live-reporting.vue';
import FieldFastActionOutput from './form/fields/fast-action-output.vue';
import FieldOpenedResolvedFilter from './form/fields/opened-resolved-filter.vue';
import FieldInfoPopup from './form/fields/info-popup.vue';
import FieldResizeColumnBehavior from './form/fields/resize-column-behavior.vue';

/**
 * Component to regroup the alarms list settings fields
 */
export default {
  name: SIDE_BARS.alarmSettings,
  components: {
    WidgetSettingsGroup,
    WidgetSettings,
    FieldTitle,
    FieldDefaultSortColumn,
    FieldColumns,
    FieldLiveReporting,
    FieldPeriodicRefresh,
    FieldDefaultElementsPerPage,
    FieldOpenedResolvedFilter,
    FieldFilters,
    FieldTextEditorWithTemplate,
    FieldSwitcher,
    FieldFastActionOutput,
    FieldGridRangeSize,
    FieldRemediationInstructionsFilters,
    FieldInfoPopup,
    FieldDensity,
    FieldNumber,
    ExportCsvForm,
    ChartsForm,
    FieldResizeColumnBehavior,
  },
  mixins: [
    widgetSettingsMixin,
    entitiesInfosMixin,
    alarmVariablesMixin,
    widgetTemplatesMixin,
    permissionsWidgetsAlarmsListFilters,
    permissionsWidgetsAlarmsListRemediationInstructionsFilters,
  ],
  computed: {
    preparedWidgetColumns() {
      return formToWidgetColumns(this.form.parameters.widgetColumns).map(column => ({
        ...column,

        text: getWidgetColumnLabel(column, ALARM_FIELDS_TO_LABELS_KEYS),
      }));
    },

    sortablePreparedWidgetColumns() {
      return this.preparedWidgetColumns.filter(column => getWidgetColumnSortable(column, ALARM_UNSORTABLE_FIELDS));
    },
  },
  mounted() {
    this.fetchInfos();
  },
  methods: {
    updateWidgetColumnsTemplate(template, columns) {
      this.$set(this.form.parameters, 'widgetColumnsTemplate', template);
      this.$set(this.form.parameters, 'widgetColumns', columns);
    },

    updateWidgetGroupColumnsTemplate(template, columns) {
      this.$set(this.form.parameters, 'widgetGroupColumnsTemplate', template);
      this.$set(this.form.parameters, 'widgetGroupColumns', columns);
    },

    updateServiceDependenciesColumnsTemplate(template, columns) {
      this.$set(this.form.parameters, 'serviceDependenciesColumnsTemplate', template);
      this.$set(this.form.parameters, 'serviceDependenciesColumns', columns);
    },

    updateMoreInfo(content, template) {
      this.$set(this.form.parameters, 'moreInfoTemplate', content);

      if (template && template !== this.form.parameters.moreInfoTemplateTemplate) {
        this.$set(this.form.parameters, 'moreInfoTemplateTemplate', template);
      }
    },
  },
};
</script>
