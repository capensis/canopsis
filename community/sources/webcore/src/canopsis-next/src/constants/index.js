export * from './recurrence-rule';
export * from './common';
export * from './entity';
export * from './alarm';
export * from './broadcast-message';
export * from './date';
export * from './dynamic-info';
export * from './event-filter';
export * from './pattern';
export * from './healthcheck';
export * from './idle-rule';
export * from './junit';
export * from './modal';
export * from './pbehavior';
export * from './permission';
export * from './remediation';
export * from './scenario';
export * from './snmp';
export * from './stats';
export * from './store';
export * from './widget';
export * from './kpi';
export * from './user';
export * from './service-weather';
export * from './map';
export * from './flowchart';
export * from './parameters';
export * from './declare-ticket';
export * from './link-rule';
export * from './external-data';
export * from './tag';
export * from './counter';
export * from './calendar';
export * from './icon';
export * from './availability';
