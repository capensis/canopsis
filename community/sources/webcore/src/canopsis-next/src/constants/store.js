export const ENTITIES_TYPES = {
  alarm: 'alarm',
  alarmDetails: 'alarmDetails',
  alarmDetailsSteps: 'alarmDetailsSteps',
  entity: 'entity',
  service: 'entityservices',
  weatherService: 'weatherService',
  pbehavior: 'pbehavior',
  event: 'event',
  userPreference: 'userPreference',
  group: 'group',
  view: 'view',
  viewTab: 'viewTab',
  widget: 'widget',
  widgetTemplate: 'widgetTemplate',
  stat: 'stat',
  user: 'user',
  role: 'role',
  eventFilter: 'eventFilter',
  metaAlarmRule: 'metaAlarmRule',
  snmpRule: 'snmpRule',
  dynamicInfo: 'dynamicInfo',
  broadcastMessage: 'broadcastMessage',
  playlist: 'playlist',
  pbehaviorExceptions: 'pbehaviorExceptions',
  pbehaviorTypes: 'pbehaviorTypes',
  pbehaviorReasons: 'pbehaviorReasons',
  remediationInstruction: 'remediationInstruction',
  remediationJob: 'remediationJob',
  remediationConfiguration: 'remediationConfiguration',
  remediationInstructionExecution: 'remediationInstructionExecution',
  remediationInstructionStats: 'remediationInstructionStats',
  scenario: 'scenario',
  entityCategory: 'entityCategory',
  testSuite: 'testSuite',
  testSuiteHistory: 'testSuiteHistory',
  stateSetting: 'stateSetting',
  idleRules: 'idleRules',
  flappingRules: 'flappingRules',
  resolveRules: 'resolveRules',
  filter: 'filter',
  ratingSettings: 'ratingSettings',
  pattern: 'pattern',
  map: 'map',
  alarmTag: 'alarmTag',
  shareToken: 'shareToken',
  declareTicketRule: 'declareTicketRule',
};
