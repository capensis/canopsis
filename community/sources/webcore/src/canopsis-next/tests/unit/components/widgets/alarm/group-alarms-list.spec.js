import flushPromises from 'flush-promises';
import Faker from 'faker';

import { mount, shallowMount, createVueInstance } from '@unit/utils/vue';
import { createMockedStoreModules } from '@unit/utils/store';
import { fakeStaticAlarms } from '@unit/data/alarm';
import { generateDefaultAlarmListWidget } from '@/helpers/entities';

import { ALARMS_OPENED_VALUES, SORT_ORDERS } from '@/constants';

import GroupAlarmsList from '@/components/widgets/alarm/group-alarms-list.vue';

jest.mock('file-saver', () => ({
  saveAs: jest.fn(),
}));

const localVue = createVueInstance();

const stubs = {
  'alarms-list-table': true,
  'c-table-pagination': true,
};

const factory = (options = {}) => shallowMount(GroupAlarmsList, {
  localVue,
  stubs,

  ...options,
});

const snapshotFactory = (options = {}) => mount(GroupAlarmsList, {
  localVue,
  stubs,

  ...options,
});

const selectTablePaginationField = wrapper => wrapper.find('c-table-pagination-stub');
const selectAlarmsListTable = wrapper => wrapper.find('alarms-list-table-stub');

describe('group-alarms-list', () => {
  const nowTimestamp = 1386435600000;

  const consequencesAlarms = fakeStaticAlarms({
    totalItems: 5,
    timestamp: nowTimestamp,
  });

  const causesAlarms = fakeStaticAlarms({
    totalItems: 7,
    timestamp: nowTimestamp,
  });

  const widget = generateDefaultAlarmListWidget();

  const defaultQuery = {
    limit: 13,
    category: 'category-id',
    opened: null,
    search: 'search',
  };
  const updateQuery = jest.fn();
  const updateLockedQuery = jest.fn();
  const queryModule = {
    name: 'query',
    getters: {
      getQueryById: () => () => defaultQuery,
    },
    actions: {
      update: updateQuery,
      updateLocked: updateLockedQuery,
    },
  };
  const tabId = 'tab-id';

  const store = createMockedStoreModules([
    queryModule,
  ]);

  afterEach(() => {
    updateLockedQuery.mockClear();
    updateQuery.mockClear();
  });

  it('Query updated after mounted', async () => {
    factory({
      store,
      propsData: {
        alarm: {},
        tabId,
        widget,
      },
    });

    await flushPromises();

    expect(updateQuery).toBeCalledTimes(1);
    expect(updateQuery).toBeCalledWith(
      expect.any(Object),
      {
        id: widget._id,
        query: {
          active_columns: widget.parameters.widgetColumns.map(v => v.value),
          limit: widget.parameters.itemsPerPage,
          opened: ALARMS_OPENED_VALUES.opened,
          with_instructions: true,
          multiSortBy: [],
          page: 1,
        },
      },
      undefined,
    );
  });

  it('Query updated after change pagination on the table', async () => {
    const wrapper = factory({
      store,
      propsData: {
        alarm: {},
        tabId,
        widget,
      },
    });

    updateQuery.mockClear();

    const alarmsListTable = selectAlarmsListTable(wrapper);

    const multiSortBy = [
      {
        sortBy: Faker.datatype.string(),
        descending: false,
      },
    ];
    const newQuery = {
      sortBy: Faker.datatype.string(),
      descending: true,
      multiSortBy,
    };

    alarmsListTable.vm.$emit('update:pagination', newQuery);

    await flushPromises();

    expect(updateQuery).toBeCalledTimes(1);
    expect(updateQuery).toBeCalledWith(
      expect.any(Object),
      {
        id: widget._id,
        query: {
          ...defaultQuery,
          sortDir: SORT_ORDERS.desc,
          sortKey: newQuery.sortBy,
          opened: ALARMS_OPENED_VALUES.all,
          multiSortBy,
        },
      },
      undefined,
    );
  });

  it('Query updated after trigger update page on the table pagination', async () => {
    const wrapper = factory({
      store,
      propsData: {
        alarm: {},
        tabId,
        widget,
      },
    });

    updateQuery.mockClear();

    const tablePagination = selectTablePaginationField(wrapper);

    const newPage = Faker.datatype.number();

    tablePagination.vm.$emit('update:page', newPage);

    await flushPromises();

    expect(updateQuery).toBeCalledWith(
      expect.any(Object),
      {
        id: widget._id,
        query: {
          ...defaultQuery,
          page: newPage,
        },
      },
      undefined,
    );
  });

  it('Query updated after trigger update records per page on the table pagination', async () => {
    const wrapper = factory({
      store,
      propsData: {
        alarm: {},
        tabId,
        widget,
      },
    });

    updateQuery.mockClear();

    const tablePagination = selectTablePaginationField(wrapper);

    const newLimit = Faker.datatype.number();

    tablePagination.vm.$emit('update:rows-per-page', newLimit);

    await flushPromises();

    expect(updateLockedQuery).toBeCalledWith(
      expect.any(Object),
      {
        id: widget._id,
        query: {
          limit: newLimit,
        },
      },
      undefined,
    );
  });

  it('Renders `group-alarms-list` with default props', () => {
    const wrapper = snapshotFactory({
      store,
      propsData: {
        alarm: {},
        tabId,
        widget,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('Renders `group-alarms-list` with consequences alarms', () => {
    const wrapper = snapshotFactory({
      store,
      propsData: {
        alarm: {
          consequences: {
            data: consequencesAlarms,
          },
        },
        tabId,
        widget,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('Renders `group-alarms-list` with causes alarms', () => {
    const wrapper = snapshotFactory({
      store,
      propsData: {
        alarm: {
          causes: {
            data: causesAlarms,
          },
        },
        tabId,
        widget,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('Renders `group-alarms-list` with sort key', () => {
    const wrapper = snapshotFactory({
      store: createMockedStoreModules([
        {
          ...queryModule,
          getters: {
            getQueryById: () => () => ({
              ...defaultQuery,
              sortKey: 'name',
              sortDir: 'desc',
            }),
          },
        },
      ]),
      propsData: {
        alarm: {
          causes: {
            data: causesAlarms,
          },
        },
        tabId,
        widget,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('Renders `group-alarms-list` without group columns', () => {
    const wrapper = snapshotFactory({
      store,
      propsData: {
        alarm: {
          causes: {
            data: causesAlarms,
          },
        },
        widget: {
          ...widget,
          parameters: {
            ...widget.parameters,
            widgetGroupColumns: undefined,
          },
        },
        tabId,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
