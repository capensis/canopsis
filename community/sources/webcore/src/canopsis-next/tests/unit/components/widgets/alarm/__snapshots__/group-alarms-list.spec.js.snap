// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group-alarms-list Renders \`group-alarms-list\` with causes alarms 1`] = `
<div>
  <alarms-list-table-stub
    alarms=""
    columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    expandable="expandable"
    has-columns="true"
    hide-groups="hide-groups"
    pagination="[object Object]"
    parent-alarm="[object Object]"
    total-items="7"
    widget="[object Object]"
  />
  <c-table-pagination-stub
    rows-per-page="13"
    total-items="7"
  />
</div>
`;

exports[`group-alarms-list Renders \`group-alarms-list\` with consequences alarms 1`] = `
<div>
  <alarms-list-table-stub
    alarms=""
    columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    expandable="expandable"
    has-columns="true"
    hide-groups="hide-groups"
    is-editing-mode="true"
    pagination="[object Object]"
    parent-alarm="[object Object]"
    total-items="5"
    widget="[object Object]"
  />
  <c-table-pagination-stub
    rows-per-page="13"
    total-items="5"
  />
</div>
`;

exports[`group-alarms-list Renders \`group-alarms-list\` with default props 1`] = `
<div>
  <alarms-list-table-stub
    alarms=""
    columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    expandable="expandable"
    has-columns="true"
    hide-groups="hide-groups"
    pagination="[object Object]"
    parent-alarm="[object Object]"
    total-items="0"
    widget="[object Object]"
  />
  <c-table-pagination-stub
    rows-per-page="13"
    total-items="0"
  />
</div>
`;

exports[`group-alarms-list Renders \`group-alarms-list\` with sort key 1`] = `
<div>
  <alarms-list-table-stub
    alarms=""
    columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    expandable="expandable"
    has-columns="true"
    hide-groups="hide-groups"
    pagination="[object Object]"
    parent-alarm="[object Object]"
    total-items="7"
    widget="[object Object]"
  />
  <c-table-pagination-stub
    rows-per-page="13"
    total-items="7"
  />
</div>
`;

exports[`group-alarms-list Renders \`group-alarms-list\` without group columns 1`] = `
<div>
  <alarms-list-table-stub
    alarms=""
    columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    expandable="expandable"
    has-columns="true"
    hide-groups="hide-groups"
    pagination="[object Object]"
    parent-alarm="[object Object]"
    total-items="7"
    widget="[object Object]"
  />
  <c-table-pagination-stub
    rows-per-page="13"
    total-items="7"
  />
</div>
`;
